### PR TITLE
libx265: update to 3.5

### PR DESCRIPTION
**Pull request by jeevatattva**
Follow-up of PR https://github.com/termux/termux-packages/pull/7401

### CODE REVIEW

**Comment by jeevatattva** [comment]
https://github.com/termux/termux-packages/pull/7548/checks?check_run_id=3623789908#step:3:623

${ARM_ARGS} should've been `-march=armv6 -mfloat-abi=soft -mfpu=vfp -marm -fPIC` but instead only `-fno-integrated-as` is present on the assembler line.

File: packages/libx265/source-CMakeLists.txt.patch [hunk 1]
```diff
@@ -16,3 +17,21 @@ diff -u -r ../x265_2.2/source/CMakeLists.txt ./source/CMakeLists.txt
...
      if(LIBRT)
          list(APPEND PLATFORM_LIBS rt)
+@@ -240,7 +243,7 @@ if(GCC)
+     endif()
+     if(ARM AND CROSS_COMPILE_ARM)
+         if(ARM64)
+-            set(ARM_ARGS -fPIC)
++            set(ARM_ARGS --target=@CCTERMUX_HOST_PLATFORM@ -fPIC)
+         else()
+             set(ARM_ARGS -march=armv6 -mfloat-abi=soft -mfpu=vfp -marm -fPIC)
```
[the remark this answers is not in the document]
Comment (jeevatattva) [reply packages/libx265/source-CMakeLists.txt.patch:27]: Thanks for your suggestion, I'll update the arguments. But it worked fine with the project defaults, which seem to be missing in this case.

**Comment by Grimler91** [comment]
This recent commit might help: https://bitbucket.org/multicoreware/x265_git/commits/4bf31dc15fb6d1f93d12ecf21fad5e695f0db5c0

**Comment by jeevatattva** [comment]
> This recent commit might help: https://bitbucket.org/multicoreware/x265_git/commits/4bf31dc15fb6d1f93d12ecf21fad5e695f0db5c0

If that's the `Apple` patch, I don't think we need it now. The existing `aarch64` asm in [release 3.5](https://bitbucket.org/multicoreware/x265_git/src/Release_3.5/source/common/aarch64/asm.S) was contributed by Huawei.

**Comment by tomty89** [comment]
> https://github.com/termux/termux-packages/pull/7548/checks?check_run_id=3623789908#step:3:623
> 
> ${ARM_ARGS} should've been `-march=armv6 -mfloat-abi=soft -mfpu=vfp -marm -fPIC` but instead only `-fno-integrated-as` is present on the assembler line.

Maybe `ARM AND CROSS_COMPILE_ARM` is never true? Perhaps `ARM AND ` should taken away.

In current master (after being properly indented), for ARM64:
```
if(ARM64 OR CROSS_COMPILE_ARM64)
    set(ARM_ARGS -fPIC -flax-vector-conversions)
    find_package(Neon)
    if(CPU_HAS_NEON)
        add_definitions(-DHAVE_NEON)
    endif()
endif()
```

Yet for ARM:
```
if(ARM AND CROSS_COMPILE_ARM)
    message(STATUS "cross compile arm")
    set(ARM_ARGS -march=armv6 -mfloat-abi=soft -mfpu=vfp -marm -fPIC)
elseif(ARM)
    find_package(Neon)
    if(CPU_HAS_NEON)
        set(ARM_ARGS -mcpu=native -mfloat-abi=hard -mfpu=neon -marm -fPIC)
        add_definitions(-DHAVE_NEON)
    else()
        set(ARM_ARGS -mcpu=native -mfloat-abi=hard -mfpu=vfp -marm)
    endif()
endif()
```
Seems to make more sense if `ARM AND ` is removed.

**Comment by jeevatattva** [comment]
@tomty89 we need to set `-DCROSS_COMPILE_ARM=ON` to indicate cross-compilation, which I've done. And it seems to work for `aarch64`. Could you review my changes. 

After forcing `--target=armv7a-linux-androideabi24` for assembly code,

```
[60/73] Linking CXX shared library libx265.so                            
FAILED: libx265.so 
: && /home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/clang++ --target=armv7-none-linux-androideabi --gcc-toolchain=/home/builder/.termux-build/_cache/android-r21d-api-24-v4 --sysroot=/home/builder/.termux-build/_cache/android-r21d-api-24-v4/sysroot -fPIC -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -Oz --target=armv7a-linux-androideabi24  -I/data/data/com.termux/files/usr/include -O3 -DNDEBUG  -Wl,--fix-cortex-a8 -L/data/data/com.termux/files/usr/lib -Wl,-rpath=/data/data/com.termux/files/usr/lib -march=armv7-a -Wl,--enable-new-dtags -Wl,--as-needed -Wl,-z,relro,-z,now --target=armv7a-linux-androideabi24  -Wl,-Bsymbolic,-znoexecstack -shared -Wl,-soname,libx265.so -o libx265.so asm.S.o cpu-a.S.o mc-a.S.o sad-a.S.o pixel-util.S.o ssd-a.S.o blockcopy8.S.o ipfilter8.S.o dct-a.S.o encoder/CMakeFiles/encoder.dir/analysis.cpp.o encoder/CMakeFiles/encoder.dir/search.cpp.o encoder/CMakeFiles/encoder.dir/bitcost.cpp.o encoder/CMakeFiles/encoder.dir/motion.cpp.o encoder/CMakeFiles/encoder.dir/slicetype.cpp.o encoder/CMakeFiles/encoder.dir/frameencoder.cpp.o encoder/CMakeFiles/encoder.dir/framefilter.cpp.o encoder/CMakeFiles/encoder.dir/level.cpp.o encoder/CMakeFiles/encoder.dir/nal.cpp.o encoder/CMakeFiles/encoder.dir/sei.cpp.o encoder/CMakeFiles/encoder.dir/sao.cpp.o encoder/CMakeFiles/encoder.dir/entropy.cpp.o encoder/CMakeFiles/encoder.dir/dpb.cpp.o encoder/CMakeFiles/encoder.dir/ratecontrol.cpp.o encoder/CMakeFiles/encoder.dir/reference.cpp.o encoder/CMakeFiles/encoder.dir/encoder.cpp.o encoder/CMakeFiles/encoder.dir/api.cpp.o encoder/CMakeFiles/encoder.dir/weightPrediction.cpp.o common/CMakeFiles/common.dir/arm/asm-primitives.cpp.o common/CMakeFiles/common.dir/primitives.cpp.o common/CMakeFiles/common.dir/pixel.cpp.o common/CMakeFiles/common.dir/dct.cpp.o common/CMakeFiles/common.dir/lowpassdct.cpp.o common/CMakeFiles/common.dir/ipfilter.cpp.o common/CMakeFiles/common.dir/intrapred.cpp.o common/CMakeFiles/common.dir/loopfilter.cpp.o common/CMakeFiles/common.dir/constants.cpp.o common/CMakeFiles/common.dir/cpu.cpp.o common/CMakeFiles/common.dir/version.cpp.o common/CMakeFiles/common.dir/threading.cpp.o common/CMakeFiles/common.dir/threadpool.cpp.o common/CMakeFiles/common.dir/wavefront.cpp.o common/CMakeFiles/common.dir/md5.cpp.o common/CMakeFiles/common.dir/bitstream.cpp.o common/CMakeFiles/common.dir/yuv.cpp.o common/CMakeFiles/common.dir/shortyuv.cpp.o common/CMakeFiles/common.dir/picyuv.cpp.o common/CMakeFiles/common.dir/common.cpp.o common/CMakeFiles/common.dir/param.cpp.o common/CMakeFiles/common.dir/frame.cpp.o common/CMakeFiles/common.dir/framedata.cpp.o common/CMakeFiles/common.dir/cudata.cpp.o common/CMakeFiles/common.dir/slice.cpp.o common/CMakeFiles/common.dir/lowres.cpp.o common/CMakeFiles/common.dir/piclist.cpp.o common/CMakeFiles/common.dir/predict.cpp.o common/CMakeFiles/common.dir/scalinglist.cpp.o common/CMakeFiles/common.dir/quant.cpp.o common/CMakeFiles/common.dir/deblock.cpp.o common/CMakeFiles/common.dir/scaler.cpp.o   && :
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: error: common/CMakeFiles/common.dir/primitives.cpp.o: multiple definition of 'x265_cpu_fast_neon_mrc_test'
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: cpu-a.S.o: previous definition here
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: error: common/CMakeFiles/common.dir/primitives.cpp.o: multiple definition of 'x265_cpu_neon_test'
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: cpu-a.S.o: previous definition here
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: warning: shared library text segment is not shareable
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
```

**Comment by tomty89** [comment]
Maybe you should set `-DHAVE_NEON`.

**Comment by jeevatattva** [comment]
With the following configuration in `build.sh`,

```sh
termux_step_pre_configure() {  
        sed -i "s/@CCTERMUX_HOST_PLATFORM@/${CCTERMUX_HOST_PLATFORM}/" \
                ${TERMUX_PKG_SRCDIR}/source/CMakeLists.txt
}

termux_step_configure() {
        termux_setup_cmake
        
        cmake   -DCMAKE_TOOLCHAIN_FILE=${TERMUX_PKG_BUILDER_DIR}/${TERMUX_ARCH}-crosscompile.cmake \
                -Wno-dev \
                ${TERMUX_PKG_SRCDIR}/source \
                -DCMAKE_INSTALL_PREFIX=$TERMUX_PREFIX
}
```

Cmake crosscompile file,
```cmake
set(CROSS_COMPILE_ARM 1)
set(CMAKE_SYSTEM_NAME Linux)
set(CMAKE_SYSTEM_PROCESSOR armv6l)

# specify the cross compiler
set(CMAKE_C_COMPILER armv7a-linux-androideabi24-clang)
set(CMAKE_CXX_COMPILER armv7a-linux-androideabi24-clang++)

# specify the target environment
SET(CMAKE_FIND_ROOT_PATH  /home/builder/.termux-build/_cache/android-r21d-api-24-v4)
```

Compilation output,
```
[ 60%] Building CXX object encoder/CMakeFiles/encoder.dir/weightPrediction.cpp.o
[ 60%] Built target encoder
[ 62%] Generating ssd-a.S.o
[ 63%] Generating asm.S.o
[ 64%] Generating blockcopy8.S.o
[ 65%] Generating cpu-a.S.o
[ 67%] Generating dct-a.S.o
[ 68%] Generating ipfilter8.S.o
[ 69%] Generating mc-a.S.o
[ 70%] Generating pixel-util.S.o
[ 71%] Generating sad-a.S.o
[ 73%] Linking CXX shared library libx265.so
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: error: cannot find -lpthread
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: error: cannot find -lrt
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: warning: shared library text segment is not shareable
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/x265-shared.dir/build.make:236: libx265.so.199] Error 1
make[1]: *** [CMakeFiles/Makefile2:151: CMakeFiles/x265-shared.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

Failed to find `pthread` and `rt`!!

Changing processor type,
```cmake
set(CMAKE_SYSTEM_PROCESSOR armv7-a)
```
and ARM arguments, `-march=armv7-a -mfloat-abi=softfp -mfpu=neon -mthumb -fPIC`.

Compilation output,

```
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: error: common/CMakeFiles/common.dir/primitives.cpp.o: multiple definition of 'x265_cpu_fast_neon_mrc_test'
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: cpu-a.S.o: previous definition here
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: error: common/CMakeFiles/common.dir/primitives.cpp.o: multiple definition of 'x265_cpu_neon_test'
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: cpu-a.S.o: previous definition here
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: error: cannot find -lpthread
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: error: cannot find -lrt
/home/builder/.termux-build/_cache/android-r21d-api-24-v4/bin/../lib/gcc/arm-linux-androideabi/4.9.x/../../../../arm-linux-androideabi/bin/ld.real: warning: shared library text segment is not shareable
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [CMakeFiles/x265-shared.dir/build.make:236: libx265.so.199] Error 1
make[1]: *** [CMakeFiles/Makefile2:151: CMakeFiles/x265-shared.dir/all] Error 2
make: *** [Makefile:136: all] Error 2
```

Doesn't seem to work with `armv7-a`.

I remember compiling the `arm` version without errors but now `pthread` and `rt, obsolete anyway,` seem to be missing.

**Comment by jeevatattva** [comment]
Need fresh eyes to fix `arm` build. 

@Grimler91 The speed of apple patch is quite considerable compared to release 3.5. I'm currently, using latest commit from x265_git master branch with ffmpeg, and I don't need to use `cpulimit` anymore. 👍

**Comment by jeevatattva** [comment]
Closing in favor of https://github.com/termux/termux-packages/pull/7561.

**Pull request closed without merging**